src/bp-forums/groups.php

```diff
@@ -1390,6 +1494,119 @@ public function replies_pagination( $args ) {
 			return $args;
 		}
 
+		/**
+		 * Ensure that forum content associated with a BuddyBoss group can only be viewed via the group URL.
+		 *
+		 * @since BuddyBoss 1.7.6
+		 *
+		 * @return void
+		 */
+		public function forum_redirect_canonical() {
+			$group_id = bp_get_current_group_id();
+
+			if ( empty( $group_id ) || ! bp_is_current_action( $this->slug ) ) {
+				return;
+			}
+
+			$forum_ids = bbp_get_group_forum_ids( $group_id );
+			$forum_id  = empty( $forum_ids ) ? false : current( $forum_ids );
+
+			if ( empty( $forum_id ) ) {
+				return;
+			}
+
+			// When navigate to group forum.
+			if ( ! bp_is_group_forum_topic() ) {
+				$group      = groups_get_group( array( 'group_id' => $group_id ) );
+				$group_link = trailingslashit( bp_get_group_permalink( $group ) );
+				$query_page = get_query_var( 'paged' );
+				$page       = empty( $query_page ) ? '' : 'page/' . $query_page;
+				$actions    = bp_action_variables();
+
+				if ( empty( $actions ) ) {
+					$redirect_to = trailingslashit( $group_link . $this->slug . '/' . get_page_uri( $forum_id ) );
+
+					// Redirect to the first forum when action variables is empty.
+					bp_core_redirect( $redirect_to );
+				}
+
+				$query = new WP_Query(
+					array(
+						'name'      => get_query_var( 'name' ),
+						'post_type' => bbp_get_forum_post_type(),
+						'orderby'   => 'ID',
+						'order'     => 'ASC',
+					)
+				);
+
+				if ( empty( $query->post ) ) {
+					return;
+				}
+
+				$child_forum = $query->post;
+				$uri         = $this->page_uri();
+				$forum       = get_page_by_path( $uri, 'OBJECT', bbp_get_forum_post_type() );
+
+				if ( empty( $forum->ID ) ) {
+					$uri         = get_page_uri( $child_forum );
+					$redirect_to = trailingslashit( $group_link . $this->slug . '/' . $uri . '/' . $page );
+
+					bp_core_redirect( $redirect_to );
+				}
+
+				$group_ids      = bb_get_child_forum_group_ids( $forum->ID );
+				$this->forum_id = ! empty( $group_ids ) && in_array( $group_id, $group_ids, true ) ? $forum->ID : false;
+			}
+
+			if ( bp_is_group_forum_topic() ) {
+				$query = new WP_Query(
+					array(
+						'name'      => bp_action_variable( 1 ),
+						'post_type' => bbp_get_topic_post_type(),
+						'orderby'   => 'ID',
+						'order'     => 'ASC',
+					)
+				);
+
+				$topic = $query->post;
+
+				if ( empty( $topic ) ) {
+					return;
+				}
+
+				$topic_forum_id = bbp_get_topic_forum_id( $topic->ID );
+				$group_ids      = bb_get_child_forum_group_ids( $topic_forum_id );
+				$this->forum_id = ! empty( $group_ids ) && in_array( $group_id, $group_ids, true ) ? $topic_forum_id : false;
+			}
+		}
+
+		/**
+		 * Get forum page uri from action variables.
+		 *
+		 * @since BuddyBoss 1.7.6
+		 *
+		 * @uses bp_action_variables() URL query parameter.
+		 *
+		 * @return string
+		 */
+		public function page_uri() {
```

Comment:
move this function to the end of the file.

Reply:
Move the at the end of the file.
https://github.com/buddyboss/buddyboss-platform/pull/2117/commits/b2c663fb81f0630d02690c0ac5064d62ff9f8387